Don't wipe out discovered desktops when LDAP dial fails

zmb3 opened the pull request:
Changelog: fixed a bug that could cause Windows desktops discovered via LDAP to be removed in error.

Changelog: fixed an issue that could cause failed Active Directory user lookups to cache the error rather than retry.

zmb3 committed:
Don't wipe out discovered desktops when LDAP dial fails

zmb3 commented:
@rhammonds-teleport @probakowski @danielashare WDYT about this?

I don't know that we should be returning the last known result _forever_, but we definitely can't let one failed dial cause all discovered desktops to be removed.

rhammonds-teleport commented:
> @rhammonds-teleport @probakowski @danielashare WDYT about this?
> 
> I don't know that we should be returning the last known result _forever_, but we definitely can't let one failed dial cause all discovered desktops to be removed.

Seems reasonable to me.

danielashare reviewed (approved):
I suppose worst case scenario is that LDAP queries fail constantly after an initial successful discovery and old desktops are still shown. Even in that case, a customer would be more upset that that they can't connect to their desktops (if sid not cached), rather than old resources being present.

zmb3 committed:
Improve LDAP discovery

- Set the discovered item's TTL based on the discovery interval.
  This ensures that setting a large discovery interval still works
  and doesn't result in desktops expiring before the next discovery
  attempt.
- Query the local cache (backend) for current desktops, rather than
  relying on the last discovery results.

zmb3 committed:
SID cache: don't cache error state
